### PR TITLE
feat: auth service

### PR DESCRIPTION
**Pull request by felixleo22**

### CODE REVIEW

File: src/auth/routers/manage.js [hunk 1]
```diff
@@ -139,4 +284,31 @@ router.put('/key/delete', checkAuth, async (req, res, next) => {
   return res.status(200).json({ id });
 });
 
+/**
+ * delete all apikey
+ */
+router.delete('/all', checkAuth, async (req, res, next) => {
+  await redisClient.flushall();
```
Comment (nojhamster) [inline src/auth/routers/manage.js:291]:
```suggestion
  try {
    await redisClient.flushall();
  } catch (err) {
    return next(err);
  }
```

File: src/auth/routers/manage.js [hunk 2]
```diff
@@ -139,4 +284,31 @@ router.put('/key/delete', checkAuth, async (req, res, next) => {
...
+  return res.status(200).json({ message: 'flushall' });
+});
+
+/**
+ * load apikey
+ */
+router.post('/load', checkAuth, async (req, res, next) => {
+  await load();
```
Comment (nojhamster) [inline src/auth/routers/manage.js:299]:
```suggestion
  try {
    await load();
  } catch (err) {
    return next(err);
  }
```

**Comment by felixleo22** [comment]
You can merge, i have done all of your suggestion

File: src/apikey/lib/redis.js [hunk 3]
```diff
@@ -11,29 +12,40 @@ if (!fs.pathExists('../apikey.json')) {
...
   logger.error(`Error in redis ${err}`);
 });
 
-const apiKeysJSON = Object.keys(apiKeys);
+const load = async () => {
+  if (process.env.NODE_ENV === 'production') {
+    apiKeys = await fs.readFile(path.resolve(__dirname, '..', 'apikey.json'));
+  } else {
+    apiKeys = await fs.readFile(path.resolve(__dirname, '..', 'apikey-dev.json'));
+  }
```
Comment (nojhamster) [inline src/apikey/lib/redis.js:37]:
```suggestion
  const filename = process.env.NODE_ENV === 'production' ? 'apikey.json' : 'apikey-dev.json';
  apiKeys = await fs.readFile(path.resolve(__dirname, '..', filename), 'utf8');
```
Just another way to do the same thing.

File: src/apikey/lib/redis.js [hunk 4]
```diff
@@ -11,29 +12,40 @@ if (!fs.pathExists('../apikey.json')) {
...
+  const apiKeysJSON = Object.keys(apiKeys);
+
+  apiKeysJSON.forEach(async (key) => {
+    try {
+      await redisClient.set(key, `${JSON.stringify(apiKeys[key])}`);
+    } catch (err) {
+      logger.error(`Cannot load ${key} with ${JSON.stringify(apiKeys[key])} on redis`);
+      logger.error(err);
+    }
   });
```
Comment (nojhamster) [inline src/apikey/lib/redis.js:49]:
```suggestion
  await Promise.all(
    Object.entries(apiKeys).map(async ([keyId, keyValue]) => {
      try {
        await redisClient.set(keyId, `${JSON.stringify(keyValue)}`);
      } catch (err) {
        logger.error(`Cannot load ${keyId} with ${JSON.stringify(keyValue)} on redis`);
        logger.error(err);
      }
    })
  );
```
`forEach()` will not wait for the promises to be resolved, so if you want to be sure that the keys are loaded when the `load()` function resolves, you should use either `Promise.all` (to run them in parallel) or a `for loop` (to run them sequentially).

File: src/apikey/middlewares/auth.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  // TODO check in query
+  const apikey = req.get('redis-password');
+
+  if (!apikey) {
+    return res.status(401).json({ message: 'Not authorized' });
+  }
+
+  if (apikey !== config.get('redis.password')) {
+    return res.status(401).json({ message: 'Not authorized' });
+  }
```
Comment (nojhamster) [inline src/apikey/middlewares/auth.js:20]:
```suggestion
  if (apikey !== config.get('redis.password')) {
    return res.status(401).json({ message: 'Not authorized' });
  }
```
If `apikey` is not set, it won't match the password anyway. However you can also write `if (!apikey || apikey !== config.get('redis.password'))` to cover the case where `redis.password` is not set, without duplicating code.

File: src/apikey/routers/manage.js [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+  const keys = await redisClient.keys('*');
+
+  const allKeys = {};
+
+  for (let i = 0; i < keys.length; i += 1) {
+    const key = keys[i];
+    let config = await redisClient.get(key);
+    config = JSON.parse(config);
+    allKeys[key] = config;
+  }
```
Comment (nojhamster) [inline src/apikey/routers/manage.js:124]: Don't forget to handle promise rejections, as express does not handle promises.

File: src/apikey/routers/manage.js [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+  const keys = await redisClient.keys('*');
+
+  keys.filter(async (key) => {
+    let config = await redisClient.get(key);
+    config = JSON.parse(config);
+    if (config.name === name) {
+      return res.status(403).json({ message: `Name [${name}] already exist` });
+    }
+  });
```
Comment (nojhamster) [inline src/apikey/routers/manage.js:183]: Don't forget to handle promise rejections, as express does not handle promises.

File: src/apikey/routers/manage.js [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+
+  let apikey;
+
+  try {
+    apikey = await createAuth(name, access, attributes, allowed);
+  } catch (err) {
+    return next(err);
+  }
+
+  let config = await redisClient.get(apikey);
```
Comment (nojhamster) [inline src/apikey/routers/manage.js:193]: Don't forget to handle promise rejections, as express does not handle promises.

File: src/apikey/routers/manage.js [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+    return res.status(404).json({ message: `[${apikey}] apikey doesn't exist` });
+  }
+
+  try {
+    await updateAuth(apikey, config.name, config.access, config.attributes, config.allowed);
+  } catch (err) {
+    return next(err);
+  }
+
+  let configApiKey = await redisClient.get(apikey);
```
Comment (nojhamster) [inline src/apikey/routers/manage.js:262]: Don't forget to handle promise rejections, as express does not handle promises.

File: src/apikey/routers/manage.js [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+  const apiKeysJSON = Object.keys(keys);
+
+  apiKeysJSON.forEach(async (key) => {
+    try {
+      await redisClient.set(key, `${JSON.stringify(keys[key])}`);
+    } catch (err) {
+      logger.error(`Cannot load ${key} with ${JSON.stringify(keys[key])} on redis`);
+      logger.error(err);
+    }
+  });
```
Comment (nojhamster) [inline src/apikey/routers/manage.js:336]: Same as the `load()` function in `lib/redis.js`, you should either use `Promise.all()` or a `for loop` to be sure that the keys are actually loaded when the server responds.

**Review by nojhamster — CHANGES_REQUESTED** [review]

File: src/apikey/lib/redis.js (hunk 3 above)
Comment (felixleo22) [reply src/apikey/lib/redis.js:37]: ok i follow that

**Pull request merged**